ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/JumpingGovernor.hs

```diff
@@ -0,0 +1,337 @@
+    Dynamo _ -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint state ->
+        case (state, jumpResult) of
+          (FoundIntersection, _) ->
+            -- Ignore jump results when we already found the intersection.
+            pure ()
+          (Happy, AcceptedJump goodPoint') -> do
+            -- The jump was accepted; we bump the peer's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
+            dynamoFragment <- fromJust <$> getDynamoFragment governor
+            let slicedDynamoFragment = fst $ fromJust $ splitAfterPoint dynamoFragment goodPoint'
+            chainSyncStateVar <- cschState . (Map.! peer) <$> readTVar clientHandles
+            modifyTVar chainSyncStateVar $ \chainSyncState -> chainSyncState {csCandidate = slicedDynamoFragment}
+            modifyTVar clientStates . Map.insert peer $ Jumper nextJumpVar goodPoint' Happy
+          (Happy, RejectedJump badPoint) ->
+            -- The previously happy peer just rejected a jump. We look for an
+            -- intersection.
+            lookForIntersection nextJumpVar goodPoint badPoint
+          (LookingForIntersection badPoint, AcceptedJump goodPoint') ->
+            -- The peer that was looking for an intersection accepts the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint' badPoint
+          (LookingForIntersection _, RejectedJump badPoint') ->
+            -- The peer that was looking for an intersection rejects the given
+            -- point, helping us narrow things down.
+            lookForIntersection nextJumpVar goodPoint badPoint'
+  where
+    lookForIntersection nextJumpVar goodPoint badPoint = do
+      (dynamoFragment0 :: AnchoredFragment (Header blk)) <- fromJust <$> getDynamoFragment governor -- full fragment
+      let (dynamoFragment1 :: AnchoredFragment (Header blk)) = snd $ fromJust $ splitAfterPoint dynamoFragment0 goodPoint -- fragment from goodPoint (excl)
+      let (dynamoFragment2 :: AnchoredFragment (Header blk)) = fst $ fromJust $ splitBeforePoint dynamoFragment1 badPoint -- fragment from goodPoint to badPoint (excl)
+      case (toOldestFirst dynamoFragment2 :: [Header blk]) of
+        [] -> do
+          modifyTVar clientStates . Map.insert peer $ Jumper nextJumpVar goodPoint FoundIntersection
+          electNewObjector governor
+        points -> do
+          let (middlePoint :: Header blk) = points !! (length points `div` 2)
+          writeTVar nextJumpVar $ Just $! (castPoint (blockPoint middlePoint) :: Point blk)
+          modifyTVar clientStates . Map.insert peer $ Jumper nextJumpVar goodPoint $ LookingForIntersection badPoint
+
+getDynamo ::
+  (MonadSTM m) =>
+  Handle m peer blk ->
+  STM m (Maybe peer)
+getDynamo Handle {clientStates} =
+  (fst <$>) . List.find (\(_, clientState) -> case clientState of Dynamo _ -> True; _ -> False) . Map.toList <$> readTVar clientStates
```

Comment:
That is what I did initially, mostly because that was how it was done in the happy path implementation by @amesgen. I do find it more satisfactory to see a map of all the peers, each having a state, rather than having a record with a dynamo field, an objector field, and then a map of jumper peers, but it isn't based on much. I'm happy changing this.